pkg/bytecode/value.go

```diff
@@ -197,26 +200,33 @@ func (m mapVal) Equals(v value) bool {
 	if !ok {
 		panic("internal error: Map.Equals called with non-Map value")
 	}
-	if len(m) != len(m2) {
+	if len(m.m) != len(m2.m) {
 		return false
 	}
-	for key, val := range m {
-		val2 := m2[key]
-		if val2 == nil || !val.Equals(val2) {
+	for key, val := range m.m {
+		val2, ok := m2.m[key]
+		if !ok || val2 == nil || !val.Equals(val2) {
 			return false
 		}
 	}
 	return true
 }
 
 func (m mapVal) Index(idx value) (value, error) {
-	k := idx.(stringVal)
-	key := string(k)
-	val, ok := m[key]
-	if !ok {
-		return nil, fmt.Errorf("%w %q", ErrMapKey, key)
+	switch idx := idx.(type) {
+	case stringVal:
+		val, ok := m.m[idx]
+		if !ok {
+			return nil, fmt.Errorf("%w %q", ErrMapKey, idx)
+		}
+		return val, nil
+	case numVal:
+		// special case where ranging over a map uses a numVal to
+		// access the order
+		key := m.order[int(idx)]
+		return key, nil
```

Comment:
We are a bit leery of this, but didn't see a better way to implement this - suggestions welcome!